include/up-cpp/utils/Expected.h

```diff
@@ -22,77 +22,114 @@ namespace uprotocol::utils {
-	using value_type = T;
-	using error_type = E;
-
-	constexpr Expected();
-	constexpr Expected(const Expected& other);
-	constexpr Expected(Expected&& other) noexcept;
-
-	constexpr explicit Expected(T&& v);
-	constexpr explicit Expected(E&& e);
-
-	Expected& operator=(const Expected& other);
-	Expected& operator=(Expected&& other);
-
-	Expected& operator=(const T& other);
-	Expected& operator=(T&& other);
-
-	Expected& operator=(const E& other);
-	Expected& operator=(E&& other);
-
-	constexpr explicit operator bool() const noexcept;
-	constexpr bool has_value() const noexcept;
-
-	constexpr T& value() &;
-	constexpr const T& value() const&;
-	constexpr T&& value() &&;
-	constexpr const T&& value() const&&;
-
-	constexpr const E& error() const& noexcept;
-	constexpr E& error() & noexcept;
-	constexpr const E&& error() const&& noexcept;
-	constexpr E&& error() && noexcept;
-
-	constexpr T value_or(T&& default_value) const&;
-	constexpr T value_or(T&& default_value) &&;
-
-	template <typename F>
-	constexpr auto and_then(F&& f) &;
-	template <typename F>
-	constexpr auto and_then(F&& f) const&;
-	template <typename F>
-	constexpr auto and_then(F&& f) &&;
-	template <typename F>
-	constexpr auto and_then(F&& f) const&&;
-
-	template <typename F>
-	constexpr auto or_else(F&& f) &;
-	template <typename F>
-	constexpr auto or_else(F&& f) const&;
-	template <typename F>
-	constexpr auto or_else(F&& f) &&;
-	template <typename F>
-	constexpr auto or_else(F&& f) const&&;
+public:
+	constexpr Expected() { new (&yay) T(); }
+	constexpr Expected(const T& rhs) { new (&yay) T(rhs); }
+	constexpr Expected(const Unexpected<E>& rhs) : ok(false) {
+		new (&nay) E(rhs.value());
+	}
+	template <class U = T>
+	constexpr explicit Expected(U&& rhs) {
+		new (&yay) T(std::forward<U>(rhs));
+	}
+
+	constexpr Expected(const Expected& rhs) : ok(rhs.ok) {
+		if (ok)
+			new (&yay) T(rhs.yay);
+		else
+			new (&nay) E(rhs.nay);
+	}
+
+	constexpr bool has_value() const noexcept { return ok; }
+
+	constexpr explicit operator bool() const noexcept { return ok; }
+
+	template <class U>
+	constexpr T value_or(U&& v) const& noexcept {
+		return ok ? yay : v;
+	}
+
+	constexpr T value() const {
+		if (!ok)
+			throw BadExpectedAccess(
+			    "Attempt to access value() when unexpected.");
+		return yay;
+	}
+
+	constexpr const E& error() const {
+		if (ok)
+			throw BadExpectedAccess(
+			    "Attempt to access error() when not unexpected.");
+		return nay;
+	}
+
+	constexpr T& operator*() {
+		if (!ok)
+			throw BadExpectedAccess(
+			    "Attempt to non-const dereference expected value when "
+			    "unexpected.");
+		return yay;
+	}
+
+	constexpr const T& operator*() const {
+		if (!ok)
+			throw BadExpectedAccess(
+			    "Attempt to const dereference expected value when unexpected.");
+		return yay;
+	}
+
+	constexpr T* operator->() {
+		if (!ok)
+			throw BadExpectedAccess(
+			    "Attempt to dereference expected pointer when unexpected.");
+		return &**this;
```

Comment:
Is there a more clear way to write this? Is the intent to call `operator*()` then take the address of the returned `yay`? Since it already checks `if (!ok)`, couldn't it just return the address of `yay`?

Reply:
This supports the usage pattern seen in these tests: Exception_pointer_dref_when_unexpected and ExpectComposite.
I believe this came from the Andrei Alexandrescu youtube video. It is also on the cppreference page at the top of the observers section.
It must be the case that this could be written with a std::get<T>(storage) here instead of the redirect to the operator*(). Is that the ask?